stacks-common/src/address/c32.rs

```diff
@@ -69,44 +97,34 @@ fn c32_encode(input_bytes: &[u8]) -> String {
     String::from_utf8(result).unwrap()
 }
 
-fn c32_normalize(input_str: &str) -> String {
-    let norm_str: String = input_str
-        .to_uppercase()
-        .replace("O", "0")
-        .replace("L", "1")
-        .replace("I", "1");
-    norm_str
-}
-
 fn c32_decode(input_str: &str) -> Result<Vec<u8>, Error> {
     // must be ASCII
     if !input_str.is_ascii() {
         return Err(Error::InvalidCrockford32);
     }
+    c32_decode_ascii(input_str)
+}
 
+fn c32_decode_ascii(input_str: &str) -> Result<Vec<u8>, Error> {
     let mut result = vec![];
     let mut carry: u16 = 0;
     let mut carry_bits = 0; // can be up to 5
 
-    let iter_c32_digits_opts: Vec<Option<usize>> = c32_normalize(input_str)
-        .chars()
-        .rev()
-        .map(|x| C32_CHARACTERS.find(x))
-        .collect();
-
-    let iter_c32_digits: Vec<usize> = iter_c32_digits_opts
+    let iter_c32_digits: Vec<u8> = input_str
+        .as_bytes()
         .iter()
-        .filter_map(|x| x.as_ref())
-        .map(|ref_x| *ref_x)
+        .rev()
+        .filter_map(|x| C32_CHARACTERS_MAP.get(*x as usize))
+        .filter_map(|v| u8::try_from(*v).ok())
```

Comment:
This code seems to depend on `u8::try_from(*v)` returning `None` if `*v` is `-1`.  Can this code be more explicit about this?  Such as:

```rust
.filter_map(|v| {
   if *v < 0 {
       None
   }
   else {
       // infallible due to the above check
       Some(*v as u8)
    }
})
```

Reply:
It may actually be better to use an array of `Option(u8)` rather than `-1i8` for the "null" sentinel values (this is what the b58 table does). I'll give that a try when I address the other comments and see if there's any unexpected perf issues with that approach.

Reply:
I changed the table from `[i8; 128]` to `[Option<u8>; 128]` which gets rid of this integer conversion, and the benchmark shows a slight perf increase.